src/backend/metal/src/command.rs

```diff
@@ -99,11 +95,161 @@ impl StageResources {
     }
 }
 
+enum CommandSink {
+    Immediate {
+        cmd_buffer: metal::CommandBuffer,
+        encoder_state: EncoderState,
+    },
+    Deferred {
+        passes: Vec<soft::Pass>,
+        is_encoding: bool,
+    },
+}
+
+impl CommandSink {
+    fn pre_render_commands<I>(&mut self, commands: I)
+    where
+        I: Iterator<Item = soft::RenderCommand>,
+    {
+        match *self {
+            CommandSink::Immediate { encoder_state: EncoderState::Render(ref encoder), .. } => {
+                for command in commands {
+                    exec_render(encoder, &command);
+                }
+            }
+            CommandSink::Deferred { ref mut passes, is_encoding: true } => {
```

Comment:
Is there a valid case for `Deferred` where `is_encoding` would be `false`? Just wondering whether this should be an `assert` instead.